Mobl_ios_Callisto-master/MealTrackingPilot/LogCommonMeal/Views/OccasionPicker.swift

```diff
@@ -0,0 +1,372 @@
+     
+     - parameter item: The item index that the user tapped, i.e. the Data Source index times the `infinityItemsMultiplier`.
+     */
+    func selectTappedItem(_ item: Int) {
+        delegate?.occasionPicker?(self, didTapItem: item, index: indexForItem(item))
+        selectItem(item, animated: true)
+    }
+    
+    /**
+     Configure the first item selection: If some pre-selected item was set, we select it, else we select the nearby to middle at all.
+     */
+    private func configureFirstSelection() {
+        let itemToSelect = currentSelectedItem != nil ? currentSelectedItem : Int(ceil(Float(numberOfItemsByDataSource) / 2.0))
+        selectedNearbyToMiddleItem(itemToSelect!)
+    }
+    
+    private func turnOccasionPickerOn() {
+        collectionView.isScrollEnabled = true
+    }
+    
+    private func turnOccasionPickerOff() {
+        collectionView.isScrollEnabled = false
+    }
+    
+    /**
+     This is a private helper that we use to reach the visible index of the current selected item.
+     Since we multiply the items several times to create an Infinite Scrolling experience, the index of a visible selected item may
+     not be the same as the index provided on Data Source.
+     
+     - returns: The visible index of current selected item.
+     */
+    func visibleIndexOfSelectedItem() -> Int {
+        let middleMultiplier = scrollingStyle == .infinite ? (infinityItemsMultiplier / 2) : infinityItemsMultiplier
+        let middleIndex = numberOfItemsByDataSource * middleMultiplier
+        let indexForSelectedItem: Int
+        
+        if let _ = currentSelectedItem , scrollingStyle == .default && currentSelectedItem == 0 {
+            indexForSelectedItem = 0
+        } else if let _ = currentSelectedItem {
+            indexForSelectedItem = middleIndex - (numberOfItemsByDataSource - currentSelectedItem)
+        } else {
+            let middleItem = Int(ceil(Float(numberOfItemsByDataSource) / 2.0))
+            indexForSelectedItem = middleIndex - (numberOfItemsByDataSource - middleItem)
+        }
+        
+        return indexForSelectedItem
+    }
+    
+    func selectItem(_ item : Int, animated: Bool) {
+        
+        var finalItem = item;
+        
+        if (scrollingStyle == .infinite && item < numberOfItemsByDataSource) {
+            let selectedItem = currentSelectedItem ?? Int(ceil(Float(numberOfItemsByDataSource) / 2.0))
+            let diff = (item % numberOfItemsByDataSource) - (selectedItem % numberOfItemsByDataSource)
+            finalItem = selectedItem + diff
+        }
+        
+        currentSelectedItem = finalItem % numberOfItemsByDataSource
+        delegate?.occasionPicker?(self, didSelectItem: currentSelectedItem, index: currentSelectedIndex)
+        setContentOffset(CGFloat(finalItem) * itemSpan - endCapSpan, animated: animated)
+    }
+    
+    // MARK: Scrolling Movement
+    
+    private var isMoving = false
```

Comment:
Migrate private var isMoving = false  before to // MARK: Initialization

Reply:
Done